### PR TITLE
[WIP] Enable Knative Kafka

### PR DESCRIPTION
**Pull request by aliok**
With this PR:
- KnativeKafka controller is enabled
- KnativeKafka webhook is enabled
- CSV updated
- Upstream e2e tests are run
- Downstream e2e tests are run
- Tooling (makefile, bash scripts) is updated

Missing:
- Update instructions in README about how to install/test/uninstall KnativeKafka related stuff

### CODE REVIEW

**Comment by aliok** [comment]
/test 4.5-upstream-e2e-aws-ocp-45

Haven't done any changes related to upstream e2e tests :(

File: knative-operator/pkg/apis/operator/v1alpha1/knativekafka_types.go [hunk 1]
```diff
@@ -69,3 +69,19 @@ type Channel struct {
 func init() {
 	SchemeBuilder.Register(&KnativeKafka{}, &KnativeKafkaList{})
 }
+
+// Used in s-o e2e tests. can't instantiate there because vendor dirs
+// of s-o and s-o/knative-operator are different
```
Comment (markusthoemmes) [inline knative-operator/pkg/apis/operator/v1alpha1/knativekafka_types.go:74]: Something like this should fix the dependency chain: https://github.com/openshift-knative/serverless-operator/commit/13ab6e9b3f9cab8b1f81ab52830ca6a4d5306c38. We basically need to tell go.mod to not use the internet but just look right next door for the codez.

Comment (aliok) [reply knative-operator/pkg/apis/operator/v1alpha1/knativekafka_types.go:74]: Getting this:
```
Cannot use 'metav1.ObjectMeta{ Name: name, Namespace: namespace, }' (type "github.com/openshift-knative/serverless-operator/test/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".ObjectMeta) as type "github.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".ObjectMeta 
```

It is a similar problem, but slightly different. Go doesn't see the `metav1` in `vendor` and the `meta1` in `knative-operator/vendor` as the same thing.

I didn't try hard solving this frankly.

Comment (markusthoemmes) [reply knative-operator/pkg/apis/operator/v1alpha1/knativekafka_types.go:74]: You get this on the commit I shared? Interesting, it seemed to compile fine for me... looking again.

**Comment by aliok** [comment]
4.5-operator-e2e-aws-ocp-45 gets this:
`./test/e2e-tests.sh: line 16: create_htpasswd_users: command not found`

I assume others jobs would have the same problem

File: test/lib.bash [hunk 2]
```diff
@@ -92,6 +93,9 @@ function downstream_serving_e2e_tests {
   return $failed
 }
 
+function downstream_knative_kafka_e2e_tests {
+}
```
Comment (markusthoemmes) [inline test/lib.bash:97]: This causes a syntax error (which subsequently causes the "function not found" error).

See

```
 ./test/e2e-tests.sh
./test/lib.bash: line 97: syntax error near unexpected token `}'
./test/lib.bash: line 97: `}' 
```

File: knative-operator/pkg/apis/operator/v1alpha1/knativekafka_types.go (hunk 1 above)
Comment (aliok) [reply knative-operator/pkg/apis/operator/v1alpha1/knativekafka_types.go:74]: Intellij still complains, but I think it is fine.
Created https://github.com/openshift-knative/serverless-operator/pull/577 to not interrupt the CI in this PR.

**Comment by aliok** [comment]
in `4.5-operator-e2e-aws-ocp-45`

Kafka dispatcher says [this](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-knative_serverless-operator/570/pull-ci-openshift-knative-serverless-operator-master-4.5-operator-e2e-aws-ocp-45/1316783304241844224/artifacts/operator-e2e-aws-ocp-45/operator-e2e/gather-knative/quay-io-openshift-knative-must-gather-sha256-51049ace7b41db7499a225b27bf11492e419245a3c49581020c184d6c816d410/knative-eventing/pods/kafka-ch-controller-5dd6cdb895-blqpc.log):
```
{"level":"fatal","ts":"2020-10-15T18:02:47.347Z","logger":"kafkachannel-dispatcher","caller":"dispatcher/kafkachannel.go:107","msg":"Unable to create kafka dispatcher","error":"unable to create kafka producer against Kafka bootstrap servers [my-cluster-kafka-bootstrap.kafka:9092] : kafka: client has run out of available brokers to talk to (Is your cluster reachable?)","stacktrace":"knative.dev/eventing-contrib/kafka/channel/pkg/reconciler/dispatcher.NewController\n\t/go/src/knative.dev/even
```

We're not installing Strimzi in `make test-e2e`.

**Comment by matzew** [comment]
> We're not installing Strimzi in `make test-e2e`.

As the files in the `e2e` folder currently are, the `knative_kafka_test.go` does check the deployemts, and the channel does need a cluster :-(

**Comment by aliok** [comment]
Modified `make test-e2e` to install Strimzi first. I cannot do it with a condition like
```
if [[ $TEST_KNATIVE_KAFKA == true ]]; then
...
```
because then I also need to add build tags to Knative Kafka e2e tests here: https://github.com/openshift-knative/serverless-operator/pull/570/files#diff-af612974c3cc32219a90cd1e3558b02aef85e2f7668439d00496bcbb31c333dbR16
and then use that tag in test command, ...

**Comment by aliok** [comment]
I just realized we're not creating `KnativeKafka` CR before starting the upstream e2e tests. `make install-kafka` needs to be executed by Prow.

**Comment by aliok** [comment]
https://github.com/openshift-knative/serverless-operator/pull/570/commits/6cb5bc40f9a0ef3f15d87bbe6bdf13f839c9526b now installs Strimzi and executes Kafka operator e2e tests only when TEST_KNATIVE_KAFKA is true. I had to move Kafka operator e2e tests to another package to accomplish this. Context: https://coreos.slack.com/archives/CD87JDUB0/p1602788484000700

**Comment by aliok** [comment]
> I just realized we're not creating KnativeKafka CR before starting the upstream e2e tests. make install-kafka needs to be executed by Prow.

Prow changes are not needed. I simply set `INSTALL_KAFKA` for upstream e2e tests.

**Comment by aliok** [comment]
/test 4.5-operator-e2e-aws-ocp-45

No idea what below is

```
=== Failed
=== FAIL: test/servinge2e TestKnativeVersusKubeServicesInOneNamespace (0.29s)
time="2020-10-15T20:33:07Z" level=info msg="Loading kube client config from path \"/var/run/secrets/ci.openshift.io/multi-stage/kubeconfig\""
time="2020-10-15T20:33:07Z" level=info msg="Loading kube client config from path \"/go/src/github.com/openshift-knative/serverless-operator/user1.kubeconfig\""
time="2020-10-15T20:33:07Z" level=info msg="Loading kube client config from path \"/go/src/github.com/openshift-knative/serverless-operator/user2.kubeconfig\""
```

**Comment by matzew** [comment]
because of #581 
/close

**Pull request closed without merging**